tests/unit/s2n_handshake_hashes_test.c

```diff
@@ -0,0 +1,155 @@
+                EXPECT_SUCCESS(s2n_handshake_get_hash_state_ptr(&conn, alg, &hash_state));
+                EXPECT_SUCCESS(s2n_hash_digest(hash_state, data, size));
+
+                /* Can't calculate the digest again: only one digest allowed per hash */
+                EXPECT_FAILURE_WITH_ERRNO(s2n_hash_digest(hash_state, data, size), S2N_ERR_HASH_NOT_READY);
+
+                /* Wiping the hashes allows them to be successfully reused */
+                EXPECT_OK(s2n_handshake_hashes_wipe(conn.handshake.hashes));
+                EXPECT_SUCCESS(s2n_hash_digest(hash_state, data, size));
+            }
+
+            s2n_handshake_hashes_free(&conn.handshake.hashes);
+        }
+    }
+
+    /* Test s2n_handshake_hashes_free */
+    {
+        /* Safety */
+        EXPECT_ERROR_WITH_ERRNO(s2n_handshake_hashes_free(NULL), S2N_ERR_NULL);
+
+        /* Frees the hashes with no memory leaks */
+        {
+            struct s2n_handshake_hashes *hashes = NULL;
+            EXPECT_OK(s2n_handshake_hashes_new(&hashes));
+            EXPECT_NOT_NULL(hashes);
+
+            EXPECT_OK(s2n_handshake_hashes_free(&hashes));
+            EXPECT_NULL(hashes);
+        }
+    }
+
+    /* Test s2n_handshake_hashes connection lifecycle */
+    {
+        uint8_t digest[SHA256_DIGEST_LENGTH] = { 0 };
+
+        /* A new connection's hashes are properly set up and can be used. */
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(conn);
+
+        struct s2n_hash_state *hash_state = NULL;
+        EXPECT_SUCCESS(s2n_handshake_get_hash_state_ptr(conn, S2N_HASH_SHA256, &hash_state));
+        EXPECT_NOT_NULL(hash_state);
+        EXPECT_SUCCESS(s2n_hash_digest(hash_state, digest, SHA256_DIGEST_LENGTH));
+        EXPECT_FAILURE_WITH_ERRNO(s2n_hash_digest(hash_state, digest, SHA256_DIGEST_LENGTH), S2N_ERR_HASH_NOT_READY);
+
+        /* A wiped connection's hashes can be reused. */
+        EXPECT_SUCCESS(s2n_connection_wipe(conn));
+        EXPECT_SUCCESS(s2n_handshake_get_hash_state_ptr(conn, S2N_HASH_SHA256, &hash_state));
+        EXPECT_NOT_NULL(hash_state);
+        EXPECT_SUCCESS(s2n_hash_digest(hash_state, digest, SHA256_DIGEST_LENGTH));
+
+        /* Freeing the handshake frees the hashes */
+        EXPECT_SUCCESS(s2n_connection_free_handshake(conn));
+        EXPECT_FAILURE_WITH_ERRNO(s2n_handshake_get_hash_state_ptr(conn, S2N_HASH_SHA256, &hash_state), S2N_ERR_NULL);
+        EXPECT_NULL(conn->handshake.hashes);
+
+        /* Wiping the connection should restore the freed hashes */
+        EXPECT_SUCCESS(s2n_connection_wipe(conn));
+        EXPECT_SUCCESS(s2n_handshake_get_hash_state_ptr(conn, S2N_HASH_SHA256, &hash_state));
+        EXPECT_NOT_NULL(hash_state);
+        EXPECT_SUCCESS(s2n_hash_digest(hash_state, digest, SHA256_DIGEST_LENGTH));
+
+        /* Freeing the connection should free the hashes */
+        EXPECT_SUCCESS(s2n_connection_free(conn));
```

Comment:
are you checking that the hashes are free-ed?

Reply:
Not in code: as far as I know, there's no way to do that.

We rely on CI jobs that run the unit tests using valgrind / asan to detect memory leaks.